File: solutions/ruby/rotational-cipher/1/rotational_cipher.rb
module RotationalCipher
  ALPHABET_LOWERCASE = *('a'..'z')
  ALPHABET_UPPERCASE = *('A'..'Z')
  CIPHER = ->(letter, alphabet, rot) { alphabet.include?(letter) ? alphabet[(alphabet.index(letter) + rot) % 26] : letter }

  def self.rotate(message, rot)
    message.chars.each_with_object(+'') do |character, result|
      new_character = CIPHER.call(character, ALPHABET_LOWERCASE, rot)
      result << CIPHER.call(new_character, ALPHABET_UPPERCASE, rot)
    end
  end

end

File: solutions/cpp/rotational-cipher/1/rotational_cipher.cpp
#include "rotational_cipher.h"

namespace rotational_cipher {

string rotate(string input, int rot){
    const int len = input.length();
    string result = "";
    char letter;

    for(int position = 0; position < len; position++) {
        letter = input[position];
        if (letter >= 'A' && letter <= 'Z') {
            letter = ROTATE((letter - 'A'), (char) rot) + 'A';
        }
        if (letter >= 'a' && letter <= 'z') {
            letter = ROTATE((letter - 'a'), (char) rot) + 'a';
        }
        result.push_back(letter);
    }
    return result;
}

}  // namespace rotational_cipher
